pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/CursorIntegrationTest.java

```diff
@@ -0,0 +1,432 @@
+    Map<String, String> headers = getHeaders();
+    Map<String, String> extraJsonProperties = getExtraQueryProperties();
+
+    // Get Pinot BrokerResponse without cursors
+    JsonNode pinotResponse;
+    pinotResponse =
+        ClusterTest.postQuery(pinotQuery, getBrokerQueryApiUrl(queryResourceUrl), headers, extraJsonProperties);
+    if (!pinotResponse.get("exceptions").isEmpty()) {
+      throw new RuntimeException("Got Exceptions from Query Response: " + pinotResponse);
+    }
+    int brokerResponseSize = pinotResponse.get("numRowsResultSet").asInt();
+
+    // Get a list of responses using cursors.
+    CursorResponse pinotPagingResponse;
+    pinotPagingResponse = JsonUtils.jsonNodeToObject(ClusterTest.postQuery(pinotQuery,
+        getBrokerQueryApiUrl(queryResourceUrl) + getCursorQueryProperties(_resultSize), headers,
+        getExtraQueryProperties()), CursorResponseNative.class);
+    if (!pinotPagingResponse.getExceptions().isEmpty()) {
+      throw new RuntimeException("Got Exceptions from Query Response: " + pinotPagingResponse.getExceptions().get(0));
+    }
+    List<CursorResponse> resultPages = getAllResultPages(queryResourceUrl, headers, pinotPagingResponse, _resultSize);
+
+    int brokerPagingResponseSize = 0;
+    for (CursorResponse response : resultPages) {
+      brokerPagingResponseSize += response.getNumRows();
+    }
+
+    // Compare the number of rows.
+    if (brokerResponseSize != brokerPagingResponseSize) {
+      throw new RuntimeException(
+          "Pinot # of rows from paging API " + brokerPagingResponseSize + " doesn't match # of rows from default API "
+              + brokerResponseSize);
+    }
+  }
+
+  private List<CursorResponse> getAllResultPages(String queryResourceUrl, Map<String, String> headers,
+      CursorResponse firstResponse, int numRows)
+      throws Exception {
+    numRows = numRows == 0 ? CommonConstants.CursorConfigs.DEFAULT_CURSOR_FETCH_ROWS : numRows;
+
+    List<CursorResponse> resultPages = new ArrayList<>();
+    resultPages.add(firstResponse);
+    int totalRows = firstResponse.getNumRowsResultSet();
+
+    int offset = firstResponse.getNumRows();
+    while (offset < totalRows) {
+      CursorResponse response = JsonUtils.stringToObject(ClusterTest.sendGetRequest(
+          getBrokerResponseApiUrl(queryResourceUrl, firstResponse.getRequestId()) + getCursorOffset(offset, numRows),
+          headers), CursorResponseNative.class);
+      resultPages.add(response);
+      offset += response.getNumRows();
+    }
+    return resultPages;
+  }
+
+  protected Object[][] getPageSizesAndQueryEngine() {
+    return new Object[][]{
+        {false, 2}, {false, 3}, {false, 10}, {false, 0}, //0 trigger default behaviour
+        {true, 2}, {true, 3}, {true, 10}, {true, 0} //0 trigger default behaviour
+    };
+  }
+
+  @DataProvider(name = "pageSizeAndQueryEngineProvider")
+  public Object[][] pageSizeAndQueryEngineProvider() {
+    return getPageSizesAndQueryEngine();
+  }
```

Comment:
nit: can be a single method. Also the method name and actual returned values has the order flipped (query engine / page size).